src/banach/retry.clj

```diff
@@ -4,6 +4,39 @@
    [manifold.deferred :as md]
    [manifold.time :as mt]))
 
+(defn routing
+  "Returns a strategy that routes to different strategies based on context
+  predicates. Kinda like [[cond]], but for strategies."
+  [& pred-strat-pairs]
+  (fn [d]
+    (md/chain
+     d
+     (fn [{:keys [failures] :as ctx}]
+       (if-some [strat (first (for [[pred strat] (partition 2 pred-strat-pairs)
+                                    :when (pred ctx)]
+                                strat))]
+         (strat ctx)
+         (throw (ex-info "no matching strat to route to" {})))))))
+
+(defn give-up
+  "A strategy that just gives up by reraising the most recent exception."
+  [d]
+  (md/chain d (fn [{:keys [failures]}] (throw (last failures)))))
+
+(defn fatal-ctx
+  "Creates a strategy that bails (throws the most recent exception) if the
+  context shows a fatal state (according to the given predicate)."
+  [is-fatal-ctx?]
+  (routing
+   is-fatal-ctx? give-up
+   (constantly true) identity))
```

Comment:
By the way, with this new fn you could write `up-to` as follows:

```clojure
(fatal-ctx #(-> % :failures count (>= n)))
```

... but maybe that isn't necessary :)